cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -0,0 +1,210 @@
+ * This is a wrapping service to host multiple {@link AbstractMessagingSubscriberService}s for lineage, usage
+ * and metadata subscriptions.
+ */
+public class MetadataSubscriberService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MetadataSubscriberService.class);
+
+  private static final Gson GSON = new Gson();
+
+  private final CConfiguration cConf;
+  private final RetryStrategy retryStrategy;
+  private final MetricsContext baseMetricsContext;
+  private final DatasetFramework datasetFramework;
+  private final Transactional transactional;
+  private final MultiThreadMessagingContext messagingContext;
+  private final List<Service> subscriberServices;
+  private DatasetId lineageDatasetId = LineageDataset.LINEAGE_DATASET_ID;
+
+  @Inject
+  MetadataSubscriberService(CConfiguration cConf, MessagingService messagingService,
+                            DatasetFramework datasetFramework, TransactionSystemClient txClient,
+                            MetricsCollectionService metricsCollectionService) {
+    this.cConf = cConf;
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.metadata.");
+    this.baseMetricsContext = metricsCollectionService.getContext(ImmutableMap.of(
+      Constants.Metrics.Tag.COMPONENT, Constants.Service.MASTER_SERVICES,
+      Constants.Metrics.Tag.INSTANCE_ID, "0",
+      Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace()));
+
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.datasetFramework = datasetFramework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), new TransactionSystemClientAdapter(txClient),
+        NamespaceId.SYSTEM, ImmutableMap.of(), null, null, messagingContext)),
+      org.apache.tephra.RetryStrategies.retryOnConflict(20, 100)
+    );
+
+    // Adding all metadata related subscribers
+    this.subscriberServices = Collections.singletonList(new DataAccessLineageSubscriber());
+  }
+
+  /**
+   * Sets the {@link DatasetId} for the {@link LineageDataset}. This method is only for testing.
+   */
+  @VisibleForTesting
+  public void setLineageDatasetId(DatasetId lineageDatasetId) {
+    this.lineageDatasetId = lineageDatasetId;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // Start all subscriber services. All of them has no-op in start, so they shouldn't fail.
+    Futures.successfulAsList(subscriberServices.stream().map(s -> {
+      LOG.debug("Starting metadata subscriber {}", s.getClass().getSimpleName());
+      return s.start();
+    }).collect(Collectors.toList())).get();
+
+    LOG.debug("All metadata subscribers started");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    // This never throw
```

Comment:
you mean this will not throw exception?

Reply:
The `Future` returned by `successfulAsList` would never throw an exception when `get` is call on it.

Reply:
You mean like "this never throw exception"??

Reply:
Actually the logic changed in next PR... so, no need to bother this.